File: exercises/exercise-1-testing-java/src/main/java/de/emaarco/example/application/service/bike/NotifyBikeCancelationService.java
package de.emaarco.example.application.service.bike;

import de.emaarco.example.application.port.inbound.bike.NotifyBikeCancelationUseCase;
import de.emaarco.example.application.port.outbound.bike.BikeSubscriptionRepository;
import de.emaarco.example.domain.bike.BikeSubscriptionId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class NotifyBikeCancelationService implements NotifyBikeCancelationUseCase {

    private static final Logger log = LoggerFactory.getLogger(NotifyBikeCancelationService.class);

    private final BikeSubscriptionRepository repository;

    public NotifyBikeCancelationService(BikeSubscriptionRepository repository) {
        this.repository = repository;
    }

    @Override
    public void notifyCancelation(BikeSubscriptionId subscriptionId) {
        var subscription = repository.find(subscriptionId);
        log.info("Notifying customer about cancelation: {}", subscription.customerEmail().value());
        repository.save(subscription.cancel());
    }

}

File: exercises/exercise-1-testing-kotlin/src/main/kotlin/de/emaarco/example/application/service/bike/NotifyBikeCancelationService.kt
package de.emaarco.example.application.service.bike

import de.emaarco.example.application.port.inbound.bike.NotifyBikeCancelationUseCase
import de.emaarco.example.application.port.outbound.bike.BikeSubscriptionRepository
import de.emaarco.example.domain.bike.BikeSubscriptionId
import mu.KotlinLogging
import org.springframework.stereotype.Service

@Service
class NotifyBikeCancelationService(
    private val repository: BikeSubscriptionRepository
) : NotifyBikeCancelationUseCase {

    private val log = KotlinLogging.logger {}

    override fun notifyCancelation(subscriptionId: BikeSubscriptionId) {
        val subscription = repository.find(subscriptionId)
        log.info { "Notifying customer about cancelation: ${subscription.customerEmail.value}" }
        repository.save(subscription.cancel())
    }
}
